### ISSUE TITLE
Optimizing dependencies - Nuxt 3.8.1

### ISSUE THREAD

**Issue by frasza** [nuxt/nuxt#24196]
### Environment

------------------------------
- Operating System: Windows_NT
- Node Version:     v20.9.0
- Nuxt Version:     3.8.1
- CLI Version:      3.9.1
- Nitro Version:    2.7.2
- Package Manager:  pnpm@8.9.0
- Builder:          -
- User Config:      devtools
- Runtime Modules:  -
- Build Modules:    -
------------------------------

### Reproduction

https://github.com/frasza/nuxt-optimize-deps

### Describe the bug

Since Nuxt 3.8.1 I (and few others on Discord) have noticed on development app boot that Nuxt/Vite does some dependencies optimization and it happens quite often (sometimes even on every page refresh, sometimes I noticed only once per dev server startup). Problem is that is also does full page refresh.

```ts
ℹ Vite server warmed up in 846ms                                                                                                                                                                        18:50:50
ℹ Vite client warmed up in 1195ms                                                                                                                                                                       18:50:50
✔ Nitro built in 548 ms                                                                                                                                                                           nitro 18:50:51
ℹ ✨ new dependencies optimized: @tiptap/vue-3, @tiptap/starter-kit                                                                                                                                     18:50:53
ℹ ✨ optimized dependencies changed. reloading    
```

Provided small reproduction using `tiptap` editor. On my own project I have even more of these packages that produce the issue. 
```ts
[18:43:23] ℹ ✨ new dependencies optimized: @tiptap/vue-3, @tiptap/starter-kit, @tiptap/extension-placeholder, @tiptap/extension-link, primevue/usetoast
ℹ ✨ optimized dependencies changed. reloading   

[18:43:44] ℹ ✨ new dependencies optimized: echarts/core, echarts/renderers, echarts/charts, echarts/components, vue-echarts
ℹ ✨ optimized dependencies changed. reloading 
```

It really does impact DX since its doing full page refresh when it happens. If I revert to Nuxt 3.8.0, issue is gone.

**Comment by dolbex** [comment]
I am having the same issue.

**Comment by DurtyFree** [comment]
Same issue here, this is very uncomfortable, especially on low spec developer machines

**Comment by manniL** [comment]
Please don't write "Me too" or "Same issue here" - instead leave a 👍🏻 on the issue ☺️

**Comment by nathanchase** [comment]
As far as I can tell, this dependency optimization will be triggered with any dynamically added component - particularly those introduced by third-party modules / plugins, or if you define a component using `defineAsyncComponent(() => import('[name of package]'));`

**Comment by danielroe** [comment]
I think this has been resolved in a dependency (likely vite related). With the lockfile you provided it is reproducible, but if you refresh your lockfile I can't reproduce.

**Comment by Letsion** [comment]
I am using clean Nuxt **3.8.1** (SPA) and have just (5 Minutes ago) regenerated the package-lock.json with `npx nuxi upgrade --force` and still have the problem.
I also use the feature of Nuxt to load the components dynamically.

Nuxt.config.ts:
```ts
...
components: {
        global: true,
        dirs: ['~/components'],
    },
...
 ``` 
I still have the problem every time I call a new .vue file.

**Comment by danielroe** [comment]
If you're still experiencing this, would you provide a reproduction? 🙏

**Comment by frasza** [comment]
@danielroe even with original reproduction as part of this issue, I have removed lock file, fresh pnpm install but optimizations are still happening on my side.

**Comment by danielwaltz** [comment]
@danielroe Here's a [reproduction](https://stackblitz.com/edit/nuxt-starter-xlsvcv?file=README.md) I created just now.

It includes pages using different features such as suspense and client side only. One extra thing worth noting is I needed to disable prefetching on `<NuxtLink>` in order to make the issue more apparent, as I believe prefetching caused all page dependencies to be optimized immediately.

You'll see that clicking between each of the pages causes a dependency optimization and triggers a full page reload.

**Comment by AnthonyRuelle** [comment]
Is there any progress on this problem? 
The new version has been published but we still won't be able to update.

**Comment by danielroe** [comment]
I don't believe this is a Nuxt issue but more likely a vite one that will be resolved when updating to Vite 5.

However, I can't investigate further as the reproduction doesn't reproduce the issue for me. It only prints it once into the console, and subsequent HMR and restarts of the dev server do not trigger re-optimising dependencies.

**Comment by danielwaltz** [comment]
Copy that! I will be curious to see if Vite 5 alleviate this as well.

It could also be my use case of developing inside a container making this harder to replicate, as this may be what leads to it not being a one-time optimization but an optimization every time my container is restarted.

If the reproduction I provided does optimize dependencies and reload the page on every initial route with a new dependency, for me this is what happens after every restart.

Either way, I appreciate the investigation and your hard work on Nuxt!

**Comment by zakarialounes** [comment]
> I don't believe this is a Nuxt issue but more likely a vite one that will be resolved when updating to Vite 5.
> 
> However, I can't investigate further as the reproduction doesn't reproduce the issue for me. It only prints it once into the console, and subsequent HMR and restarts of the dev server do not trigger re-optimising dependencies.

@danielroe Have you tried cloning the repository locally? You should be able to do test without cache.

**Comment by travisricks** [comment]
I don't understand this problem fully, but for what it's worth I'm also experiencing it and I do not use a container. I work locally and am getting the optimize message/app reload every time I go to a page that includes a new dependency, e.g. from Lodash.

**Comment by patroza** [comment]
please a way to disable this?

**Comment by anjarupnik** [comment]
Hi, I had the same issue. Not sure which change in nuxt@3.8.1 caused it because it didn't happen in v3.8.0, but I managed to prevent it by adding following to the nuxt.conifg:
`
vite: {
    optimizeDeps: {
      include: [add here all dependencies that are causing reloading]
    }
}
`
Those dependencies should then be optimised before server starts up. Some more links about it:
https://stackoverflow.com/questions/75469067/vite-cypress-how-to-prevent-reloading-due-to-optimized-dependencies-causin
https://vitejs.dev/config/dep-optimization-options

**Comment by zakarialounes** [comment]
> Hi, I had the same issue. Not sure which change in nuxt@3.8.1 caused it because it didn't happen in v3.8.0, but I managed to prevent it by adding following to the nuxt.conifg: `vite: { optimizeDeps: { include: [add here all dependencies that are causing reloading] } }` Those dependencies should then be optimised before server starts up. Some more links about it: https://stackoverflow.com/questions/75469067/vite-cypress-how-to-prevent-reloading-due-to-optimized-dependencies-causin https://vitejs.dev/config/dep-optimization-options

Specifying the packages present in the node_modules folder generating a reload in optimizeDeps.include corrects the problem. Having said that, I'm not sure we should add dependencies coming from node_module ourselves, since the documentation states "By default, linked packages not inside node_modules are not pre-bundled". I therefore understand that packages in the node_module folder should be pre-bundled?

**Comment by Letsion** [comment]
The error also occurs in the nuxt-nightly version. 
he error still exists here: Nuxt 3.9.0-28346301.c214869e

**Comment by anjarupnik** [comment]
@zakarialounes Yes, that is what I also understood. But it seems Vite does not find them initially so it needs to optimise them after server has started. At least that is my assumption, didn't have time to dive deeper in this issue, but hope workaround helps until this is fixed 😄

**Comment by agelinas** [comment]
I used anjarupnik's "fix" and it worked for me. In the screenprint, notice the vite optimizeDeps.
![Capture d’écran du 2023-11-28 10-53-25](https://github.com/nuxt/nuxt/assets/11081303/24f16936-cc18-4eb9-a489-f54cae049be6)

The following screenprint was taken when I had the problem which was trigerred by my browser(s). The browser page froze. 
![Capture d’écran du 2023-11-28 10-16-03](https://github.com/nuxt/nuxt/assets/11081303/c716e064-ff2e-42b2-889b-5da3d38fb8f7)

**Comment by HADB** [comment]
After some digging, I found this issue related to this problem: https://github.com/vitejs/vite/issues/5419

**Comment by gyhyfj** [comment]
Same problem, and what's worse on dev, if you're navigating to a new page, whitch imports a new dep whitch haven't been installed yet, the navigation can never be completed and have no error info in terminal.

**Comment by antfu** [comment]
I will look into it. You can try to debug this as well by passing `DEBUG=vite:deps` env to see Vite's debug log.

The thing I found is that every time Nuxt starts, the previous optimization cache is got removed:
<img width="1252" alt="image" src="https://github.com/nuxt/nuxt/assets/11247099/c8791a5f-e85b-4ab7-b4b7-2a61a412cab3">

With the insight from @patak-dev, I guess it's because the hash Vite calculated are not consistent across multiple run:
https://github.com/vitejs/vite/blob/c902545476a4e7ba044c35b568e73683758178a3/packages/vite/src/node/optimizer/index.ts#L366

I will dig and see what it causing that.

**Comment by antfu** [comment]
So, as you see on the linked PRs above, this is caused by quite a few moving parts. Basically it's because different Nuxt modules might contribute to the Vite config at a different time (racing conditions), and causing the Vite config to be unstable across runs and the hash being different: https://github.com/vitejs/vite/blob/c902545476a4e7ba044c35b568e73683758178a3/packages/vite/src/node/optimizer/index.ts#L1210-L1226

So basically, any module could cause that. The main one should be Nuxt DevTools (https://github.com/nuxt/devtools/pull/542) where the fix is already released as v1.0.6. Then we have `@nuxt/mdc` (https://github.com/nuxt-modules/mdc/pull/111) which is also a dependency of `@nuxt/content`. For other modules, we need to investigate one by one. Would love your help btw.

On the Vite side, fixes and proposals are on the way to make the experience better:
- https://github.com/vitejs/vite/pull/15337
- https://github.com/vitejs/vite/pull/15339

There is currently no userland workaround at this moment, please be patient for all those packages to be upgraded. Will keep this post updated.

**Comment by harlan-zw** [comment]
Ooh, nice find! It did feel like there was a bigger issue looming. I'll need to keep an eye out for this on my modules, the upstream PRs should help with that though.

**Comment by danielwaltz** [comment]
Awesome work! Not sure if this is the best place to track it anymore, but combing through the modules used in my project I found a place in the `vueuse` nuxt module that could be a culprit/contributing to this. 🤔 

https://github.com/vueuse/vueuse/blob/main/packages/nuxt/index.ts#L75-L79

**Comment by frasza** [comment]
Can confirm it works well with `nuxt 3.9.0` and `vue 3.4`

**Comment by travisricks** [comment]
This issue was solved on my end too. I just did a weekly package update and it's happening again. Should I file something in the Vite repo? Has anyone else had the issue happen again?

**Comment by anjarupnik** [comment]
I also updated to 3.9.3 and issue is still there but it happens less often. It seems to happen every time package-lock is updated (even if updated dependency is not the one that needs to be optimised).

**Comment by danielroe** [comment]
That's vite's intended behaviour - it will reoptimise deps when your lockfile changes.

**Comment by travisricks** [comment]
In my case, my lockfile is not changing. I'm experiencing app reloads when running `npm run dev` and navigating from one route to a new route when that new route includes a new dependency, e.g. a method from lodash.

**Comment by anjarupnik** [comment]
@danielroe I understand but I think it would be a bit nicer developer experience if they are all optimised before page is served. Currently I start the server then it reloads 3 times for each dependency and then again on route change where another package is use.

**Comment by travisricks** [comment]
> This issue was solved on my end too. I just did a weekly package update and it's happening again. Should I file something in the Vite repo? Has anyone else had the issue happen again?
> 
> I'm not exactly sure which package is causing the issue, but here are some of the packages that were updated that I think could be related.
> 
> * nuxt 3.9.1 > 3.9.3
> * vue 3.4.10 > 3.4.15
> * vite-node 1.1.3 > 1.2.1

@frasza @danielroe 
Just checking back in on this. I'm still experiencing reloads on route changes since the recent patch upgrades of the above packages.

**Comment by danielroe** [comment]
cc: @antfu

**Comment by antfu** [comment]
This issue gets a bit long and quite a lot of context involved - could anyone create a new issue linking back to this one, and with a reproduction of the latest nuxt? Otherwise I am not so sure what to "fix" - thank you!

**Comment by frasza** [comment]
@antfu https://github.com/nuxt/nuxt/issues/26783

**Comment by manniL** [comment]
I'd close here in favor of #26783 then to keep focus on the issue 👍🏻

**Comment by c0nd3v** [comment]
Still a problem @manniL

**Comment by margaretjoanmiller** [comment]
If anyone else is having this issue, try passing the dependencies that cause the reloads to vite's optimizeDeps include (https://vite.dev/config/dep-optimization-options#optimizedeps-include) in your nuxt config
